### ISSUE TITLE
DuckDB throws exception on Arrow table returned by Huggingface Datasets

### ISSUE THREAD

**Issue by srobertjames** [duckdb/duckdb#4242]
### What happens?

[Huggingface Datasets](https://huggingface.co/docs/datasets/index) is a [wrapper around Arrow](https://huggingface.co/docs/datasets/about_arrow), and exposes the underlying Arrow table via `[Dataset.data](url)`.  I'd like to use DuckDB on it, but, when I do, I get exceptions: `Python Object MemoryMappedTable not suitable for replacement scans`  or `Python Object Type MemoryMappedTable is not an accepted Arrow Object.` (depending on the call).

I'm not sure if this is a bug in DuckDB or I'm simply using it incorrectly, but my goal is to use DuckDB to do SQL on the Arrow tables from Huggingface's Dataset.

### To Reproduce

```
import duckdb
from datasets import Dataset

cursor = duckdb.connect()

ds = Dataset.from_csv('myfile.csv')
arr = ds.data
print(cursor.execute('SELECT * FROM arr LIMIT 3').fetchall())
# RuntimeError: Python Object MemoryMappedTable not suitable for replacement scans
duckdb.arrow(arr)
# RuntimeError: Python Object Type MemoryMappedTable is not an accepted Arrow Object
```

Routing through Pandas works fine:
```
p = ds.to_pandas()
print(cursor.execute('SELECT * FROM p LIMIT 3').fetchall())
# [(1, 0, 3, 'Braun' ...
```

### OS:

Linux 3de821bd7d75 5.10.123+ #1 SMP Sat Jul 16 14:54:12 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### DuckDB Version:

0.40

### DuckDB Client:

Python via Jupyter notebook

### Full Name:

S. Robert James

### Affiliation:

individual - personal project

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by Mause** [comment]
It looks like we're presently being quite strict on how we do the type checking for the replacement scans: https://github.com/duckdb/duckdb/blob/d883bc8a0563ffb378ebc320aefa2e5bc5169763/tools/pythonpkg/src/pyconnection.cpp#L671-L677

Looking at the code for the datasets library, it doesn't subclass anything we can use to check that it fulfills an interface either, so we'd probably need to check that it implements an "arrow table" Protocol

In short, this looks fixable, just requires more investigation to see what's required to fix it

**Comment by srobertjames** [comment]
`ds.data.table` works! Thank you.

**Comment by pdet** [comment]
@Mause what you mean with checking if the protocol is implemented?

I think we could support ```MemoryMappedTable``` by making the ```ds.data.table``` call internally

**Comment by Mause** [comment]
@pdet this is what I'm referring to by type protocols: https://mypy.readthedocs.io/en/stable/protocols.html

Personally I think this is the better path going forward, rather than trying to support each library individually - if a library exposed an interface that [quacks](https://realpython.com/lessons/duck-typing/) like an arrow table, we should accept it. This would be pretty vastly preferable than checking for the name of the type, which is what we're currently doing (see above)

**Comment by pdet** [comment]
Sorry for my naive question, but how should that be done in practice?
Attribute checking on the object? Or maybe there is a cleaner solution?

**Comment by Mause** [comment]
> Sorry for my naive question, but how should that be done in practice?
> Attribute checking on the object? Or maybe there is a cleaner solution?

If you enable runtime checking on the protocol, you can just use `isinstance`. Feel free to message on discord if that's easier btw
